src/app/AttributePathExpandIterator.cpp

```diff
@@ -0,0 +1,193 @@
+    {
+        mBeginAttributeIndex = 0;
+        mEndAttributeIndex   = emberAfGetServerAttributeCount(aEndpointId, aClusterId);
+    }
+    else
+    {
+        mBeginAttributeIndex = emberAfGetServerAttributeIndexByAttributeId(aEndpointId, aClusterId, aClusterInfo.mFieldId);
+        // If the given attribute id does not exist on the given endpoint, it will return uint16(0xFFFF), then endAttributeIndex
+        // will be 0
+        mEndAttributeIndex = static_cast<uint16_t>(mBeginAttributeIndex + 1);
+    }
+}
+
+bool AttributePathExpandIterator::Proceed()
+{
+    for (; mpClusterInfo != nullptr; (mpClusterInfo = mpClusterInfo->mpNext, mEndpointIndex = UINT16_MAX))
+    {
+        // Special case: If this is a concrete path, we just return its value as-is.
+        if (!mpClusterInfo->HasWildcard())
+        {
+            mOutputPath.mEndpointId  = mpClusterInfo->mEndpointId;
+            mOutputPath.mClusterId   = mpClusterInfo->mClusterId;
+            mOutputPath.mAttributeId = mpClusterInfo->mFieldId;
+
+            // Prepare for next iteration
+            (mpClusterInfo = mpClusterInfo->mpNext, mEndpointIndex = UINT16_MAX);
+            return true;
+        }
+
+        if (mEndpointIndex == UINT16_MAX)
+        {
+            PrepareEndpointIndexRange(*mpClusterInfo);
+            // If we have not started iterating over the endpoints yet.
+            mEndpointIndex = mBeginEndpointIndex;
+            mClusterIndex  = UINT8_MAX;
+        }
+
+        for (; mEndpointIndex < mEndEndpointIndex; (mEndpointIndex++, mClusterIndex = UINT8_MAX, mAttributeIndex = UINT16_MAX))
+        {
+            EndpointId endpointId = emberAfEndpointFromIndex(mEndpointIndex);
+
+            if (mClusterIndex == UINT8_MAX)
+            {
+                PrepareClusterIndexRange(*mpClusterInfo, endpointId);
+                // If we have not started iterating over the clusters yet.
+                mClusterIndex   = mBeginClusterIndex;
+                mAttributeIndex = UINT16_MAX;
+            }
+
+            for (; mClusterIndex < mEndClusterIndex; (mClusterIndex++, mAttributeIndex = UINT16_MAX))
+            {
+                // emberAfGetNthClusterId must return a valid cluster id here since we have verified the mClusterIndex does
+                // not exceed the endAttributeIndex.
+                ClusterId clusterId = emberAfGetNthClusterId(endpointId, mClusterIndex, true /* server */).Value();
+                if (mAttributeIndex == UINT16_MAX)
+                {
+                    PrepareAttributeIndexRange(*mpClusterInfo, endpointId, clusterId);
+                    // If we have not started iterating over the attributes yet.
+                    mAttributeIndex = mBeginAttributeIndex;
+                }
+
+                if (mAttributeIndex < mEndAttributeIndex)
+                {
+                    // GetServerAttributeIdByIdex must return a valid attribute here since we have verified the mAttributeindex does
```

Comment:
```suggestion
                    // GetServerAttributeIdByIdex must return a valid attribute here since we have verified the mAttributeIndex does
```